tripal/tripal
# Content Types and Fields defined by YAML

# Summary
Issue #1393 


### Tripal Version:  4

## Description


This PR moves creation of Tripal content types out of the Tripal Chado prepare step and into Services that use YAML files. It partially addresses items in Issue #1393 .


### Changes
The following changes were made:
1.  Two new YAML files for content types and fields: 
     - `tripal_chado/config/install/tripal.tripal_content_types.chado_content_types.yml`
     - `tripal_chado/config/install/tripal.tripal_fields.chado_fields.yml`
2. The schema for these configuration files have been aded to the `tripal/config/schema/tripal.schema.yml` file.
3. New configuration entity classes were added. This allows the site admin to reload the default configuration settings, which probably won't ever need to happen but is useful for when we code new content types.
   - `tripal/src/Entity/TripalContentTypes.php`
   - `tripal/src/Entity/TripalFields.php`
5. New Service classes were created that allow for creation of content types and attaching their fields:
    - `tripal/src/Services/TripalContentTypes.php`
    - `tripal/src/Services/TripalFields.php`
6.  Functions for creating content types and attaching fields were removed from the Chado Prepare task and instead, these new service classes are used to create the content types from the YAML files. 
7. A new `tripal/api/tripal.api.php` file was created, and the function `tripal_load_configuration()` was added because we need a way to reload tripal provided configurations that may be present in extension module.   This function gets called in the `tripal_rebuild()` function which gets called anytime the cache is cleared for the site.
8. All the necessary interface classes, form classes and routing/links were added.
9. Content type machine names are now more human readable. In Tripal v3 they used to be the term namespace and accession.  The field names are also human readable now.  This makes them so much easier to use as string tokens.
10. Content type URLs and titles are now  set in the YAML file (but can still be changed in the interface).  This makes it easier to set a default by the content type creator.
      - FYI... During testing, I fixed a bug in the additional types field because it wasn't giving the infrasepcific type for the organism content.
12.  I updated the descriptions for each content type.  In Tripal v3 and until now, only the controlled vocabulary term description was used, but that wasn't really well tailored for the content type page listing.  The content type developers can now set this in the YAML file too.

### More to Do
This PR does not do the following suggested in issue #1391 :
- Support indicating the `chado` schema that the field should support.  I think we should not let the user specify that in the YAML... we'll have to rethink how to do that.
- Does not put the creation of the content types and fields into the install of the module. That will be in a future PR. This PR is just step one of the process of making the YAML files work.

### Functional Tests
This  PR is a placeholder because I'm having a hard time with funcitonal testing on my machine at the moment. At least this way I can see what's wrong with and try to fix it. So, feel free to do a funcitonal review, but the code review should wait until testing passes.  

## Testing?



1. To test content type creation
    - Start with a brand new installation of Tripal
    - Install Chado and run the Prepare step. You should see that each content type is created, then afterwrads each field that gets added to a content type is listed. 
    - Go to the Admin > Manage > Structure >Tripal Content Types. You should see all of the content types listed.
    - Click the `Organism` content type. You can see it still works and the page titles and url alias options have been updated to use the new field names from the fields YAML file.
    - You should be able to see the fields attached to the content type as well by going back a page and clicking the 'manage fields' link.
    - Try creating a new organism (or other content type) page. Go to Admin > Tripal Content and click the "Add Tripal Content" button.
       -  All of the content type descriptions have been tailored.
       -  Follow through with creation of a conteont type and it should still work.
2. To test configuration reload
    - You can refresh the YAML configurations (in case you made  change) by going to `Admin` > `Tripal` > `more` (in the header). Then click `Configuration` 
    - Pick any of the three configuration types: content terms, content types or field.  A very simple listing appears.  This is where all of the configuration files for each Tripal extension module will appear.  You can delete any configuration. If you then clear the Drupal cache with a `drush cr` then the configuration will return.  Normally, this is not common Drupal behavior. If you delete a configuration there is no way to get it back.  If you are a developer, this is really annoying because it means you hae to reinstall the entire site.  This feature will most likely only be used by developers.
